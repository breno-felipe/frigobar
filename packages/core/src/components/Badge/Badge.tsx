import React from 'react';
import styled from 'styled-components';

const badgeAlignment = {
  'top-left': `
    top: 0;
    left: 0;
    transform: translate(-50%, -50%);
  `,
  'top-right': `
    top: 0;
    right: 0;
    transform: translate(50%, -50%);
  `,
  'bottom-left': `
    bottom: 0;
    left: 0;
    transform: translate(-50%, 50%);
  `,
  'bottom-right': `
    bottom: 0;
    right: 0;
    transform: translate(50%, 50%);
  `,
};

const Root = styled.span`
  display: inline-flex;
  position: relative;
`;

const StyledBadge = styled.span<{
  alignment: BadgeProps['alignment'];
  color: BadgeProps['color'];
  hasChildren: boolean;
}>(
  ({
    alignment,
    color,
    hasChildren,
    theme: { colors, spacings, radius, baseFontSize },
  }) => `
    display: flex;
    align-items: center;
    justify-content: center;

    padding: ${spacings.xxsmall}px;

    border-radius: ${radius[1]}px;
    background-color: ${(colors[color] && colors[color][500]) || color};

    color: ${colors.white};
    font-size: ${baseFontSize * 0.75}px;
    line-height: ${baseFontSize * 0.75}px;
    ${
      hasChildren
        ? `
            position: absolute;
            ${badgeAlignment[alignment]}
          `
        : ''
    }
  `,
);

interface BadgeProps {
  /** where badge will be displayed */
  children?: React.ReactNode;
  /** change badge color, it can be any color theme property, like "primary", "info", "success", etc. Or a custom name/hex color (#000) */
  color:
    | 'primary'
    | 'secondary'
    | 'info'
    | 'success'
    | 'warning'
    | 'danger'
    | 'neutral';
  /** the number to be displayed inside badge */
  content: number;
  /** in which component corner to align the badge */
  alignment: 'top-left' | 'top-right' | 'bottom-left' | 'bottom-right';
}

const Badge = ({
  children,
  color,
  content,
  alignment,
}: BadgeProps): JSX.Element => (
  <Root>
    {children}
    <StyledBadge
      color={color}
      role="status"
      aria-live="polite"
      aria-atomic="true"
      hasChildren={Boolean(React.Children.count(children))}
      alignment={alignment}
    >
      {content}
    </StyledBadge>
  </Root>
);

Badge.defaultProps = {
  color: 'primary',
  children: undefined,
  content: undefined,
  alignment: 'top-right',
};

export default Badge;
