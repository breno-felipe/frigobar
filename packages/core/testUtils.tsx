import React from 'react';
import PropTypes from 'prop-types';
import {
  render,
  fireEvent,
  waitFor,
  RenderOptions,
  RenderResult,
} from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import theme from './src/theme';

const Providers = ({ children }: { children?: React.ReactNode }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

Providers.propTypes = {
  children: PropTypes.node.isRequired,
};

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
): RenderResult => render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render, fireEvent, waitFor };
