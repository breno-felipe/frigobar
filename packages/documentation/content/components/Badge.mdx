---
title: Badge
menu: components
---

Badge component

```jsx react-live=true
<Badge color="danger" content={10}>
  <Button>Badge</Button>
</Badge>
```
