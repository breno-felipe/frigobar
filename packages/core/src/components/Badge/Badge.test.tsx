import React from 'react';
import { render } from '../../../testUtils';

import Badge from './Badge';

describe('<Badge />', () => {
  describe('Snapshots', () => {
    it('should match snapshot with default badge', () => {
      const { container } = render(<Badge content={10} />);
      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with children', () => {
      const { container } = render(
        <Badge content={10}>
          <button type="button">badge</button>
        </Badge>,
      );
      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with different color', () => {
      const { container } = render(<Badge content={10} color="success" />);
      expect(container).toMatchSnapshot();
    });

    describe('alignments', () => {
      it('should match snapshot with top-left alignment', () => {
        const { container } = render(
          <Badge content={10} alignment="top-left" />,
        );
        expect(container).toMatchSnapshot();
      });

      it('should match snapshot with top-right alignment', () => {
        const { container } = render(
          <Badge content={10} alignment="top-right" />,
        );
        expect(container).toMatchSnapshot();
      });

      it('should match snapshot with bottom-left alignment', () => {
        const { container } = render(
          <Badge content={10} alignment="bottom-left" />,
        );
        expect(container).toMatchSnapshot();
      });

      it('should match snapshot with bottom-right alignment', () => {
        const { container } = render(
          <Badge content={10} alignment="bottom-right" />,
        );
        expect(container).toMatchSnapshot();
      });
    });
  });
});
