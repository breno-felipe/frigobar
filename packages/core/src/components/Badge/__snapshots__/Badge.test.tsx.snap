// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Badge /> Snapshots alignments should match snapshot with bottom-left alignment 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 4px;
  border-radius: 4px;
  background-color: #008d96;
  color: #fff;
  font-size: 12px;
  line-height: 12px;
}

<div>
  <span
    class="c0"
  >
    <span
      aria-atomic="true"
      aria-live="polite"
      class="c1"
      color="primary"
      role="status"
    >
      10
    </span>
  </span>
</div>
`;

exports[`<Badge /> Snapshots alignments should match snapshot with bottom-right alignment 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 4px;
  border-radius: 4px;
  background-color: #008d96;
  color: #fff;
  font-size: 12px;
  line-height: 12px;
}

<div>
  <span
    class="c0"
  >
    <span
      aria-atomic="true"
      aria-live="polite"
      class="c1"
      color="primary"
      role="status"
    >
      10
    </span>
  </span>
</div>
`;

exports[`<Badge /> Snapshots alignments should match snapshot with top-left alignment 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 4px;
  border-radius: 4px;
  background-color: #008d96;
  color: #fff;
  font-size: 12px;
  line-height: 12px;
}

<div>
  <span
    class="c0"
  >
    <span
      aria-atomic="true"
      aria-live="polite"
      class="c1"
      color="primary"
      role="status"
    >
      10
    </span>
  </span>
</div>
`;

exports[`<Badge /> Snapshots alignments should match snapshot with top-right alignment 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 4px;
  border-radius: 4px;
  background-color: #008d96;
  color: #fff;
  font-size: 12px;
  line-height: 12px;
}

<div>
  <span
    class="c0"
  >
    <span
      aria-atomic="true"
      aria-live="polite"
      class="c1"
      color="primary"
      role="status"
    >
      10
    </span>
  </span>
</div>
`;

exports[`<Badge /> Snapshots should match snapshot with children 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 4px;
  border-radius: 4px;
  background-color: #008d96;
  color: #fff;
  font-size: 12px;
  line-height: 12px;
  position: absolute;
  top: 0;
  right: 0;
  -webkit-transform: translate(50%,-50%);
  -ms-transform: translate(50%,-50%);
  transform: translate(50%,-50%);
}

<div>
  <span
    class="c0"
  >
    <button
      type="button"
    >
      badge
    </button>
    <span
      aria-atomic="true"
      aria-live="polite"
      class="c1"
      color="primary"
      role="status"
    >
      10
    </span>
  </span>
</div>
`;

exports[`<Badge /> Snapshots should match snapshot with default badge 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 4px;
  border-radius: 4px;
  background-color: #008d96;
  color: #fff;
  font-size: 12px;
  line-height: 12px;
}

<div>
  <span
    class="c0"
  >
    <span
      aria-atomic="true"
      aria-live="polite"
      class="c1"
      color="primary"
      role="status"
    >
      10
    </span>
  </span>
</div>
`;

exports[`<Badge /> Snapshots should match snapshot with different color 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 4px;
  border-radius: 4px;
  background-color: #50a47f;
  color: #fff;
  font-size: 12px;
  line-height: 12px;
}

<div>
  <span
    class="c0"
  >
    <span
      aria-atomic="true"
      aria-live="polite"
      class="c1"
      color="success"
      role="status"
    >
      10
    </span>
  </span>
</div>
`;
